docs/Migration-Guide-V3.md

```diff
@@ -0,0 +1,100 @@
+# V3 Migration Guide
+
+This guide is aimed to help migrating from Fastify v2 to v3.
+
+Before beginning please ensure that any deprecation warnings from v2 are fixed as we have removed all deprecated features and they will no longer work after upgrading. (#1750)
+
+## Breaking changes
+
+### Changed middleware support (#2014)
+
+From Fastify v3, middleware support does not come out of the box with the framework itself.
+
+If you use Express middleware in your application, please install and register the [`fastify-express`](https://github.com/fastify/fastify-express) or [`middie`](https://github.com/fastify/middie) plugin before doing so.
+
+**v2:**
+
+```js
+fastify.use(require("cors")());
+```
+
+**v3:**
+
+```js
+await fastify.register(require("fastify-express"));
+fastify.use(require("cors")());
+```
+
+### Changed schema substitution (#2023)
+
+We have dropped support for non-standard `replace-way` shared schema substitution and replaced it with standard compliant `$ref-way` references.
+
+**v2:**
+
+```js
+const schema = {
+  body: "schemaId#"
+};
+fastify.route({ method, url, schema, handler });
+```
+
+**v3:**
+
+```js
+const schema = {
+  body: {
+    $ref: "schemaId#"
+  }
+};
+fastify.route({ method, url, schema, handler });
+```
+
+### Changed schema validation options (#2023)
+
+We have replaced `setSchemaCompiler` and `setSchemaResolver` options with `setValidatorCompiler` to enable future tooling improvements.
+
+**v2:**
+
+```js
+const fastify = Fastify();
+const ajv = new AJV();
+ajv.addSchema(fastClone(schemaA));
+ajv.addSchema(fastClone(schemaB));
+
+fastify.setSchemaCompiler(schema => ajv.compile(schema));
+fastify.setSchemaResolver(ref => ajv.getSchema(ref).schema);
+```
+
+**v3:**
+
+```js
+const fastify = Fastify();
+const ajv = new AJV();
+ajv.addSchema(schemaA);
+ajv.addSchema(schemaB);
+
+fastify.setValidatorCompiler((method, url, httpPart, schema) =>
+  ajv.compile(schema)
+);
+```
+
+### Changed hooks behaviour (#2004)
+
+From Fastify v3, the behavior of `onRoute` and `onRegister` hooks will change slightly in order to support hook encapsulation.
+
+- `onRoute` - The hook will be called asynchronously, in v1/v2 it's called as soon as a route is registered. This means that if you want to use it, you should register this hook as soon as possible in your code.
```

Comment:
I don't think there is `onRoute` in v1.

Reply:
I've just realised that the documentation for this needs to be updated in `Hooks.md` - It's still talking about onRoute being trigger synchronously:
```
### onRoute
Triggered when a new route is registered. Listeners are passed a `routeOptions` object as the sole parameter. The interface is synchronous, and, as such, the listeners do not get passed a callback. This hook is encapsulated.
```

Reply:
Would you mind to send a PR to update the docs?

Reply:
I think I misunderstood the sentence. What the author meant is probably that the hook runs synchronously (fire and forget) which isn't a contradiction to it being triggered asynchronously during route registration. 

I'm not clear exactly what changed though and what users need to be aware of here. 
@Eomm do you have more context of this change? Are the docs correct or can you clarify the change so I can update it?

Reply:
> I think I misunderstood the sentence. What the author meant is probably that the hook runs synchronously (fire and forget) which isn't a contradiction to it being triggered asynchronously during route registration.

Exactly.

---

The key difference between v2 and v3 in this case is that onRoute hooks will be called only for routes nested in that plugin and descendants. In case of hooks defined in something wrapped in `fastify-plugin`, the hook will be registered in its current context. 

---

The difference is